File: school-space-client/src/models/Course.ts
import {Teacher} from "./Teacher";

export class Course {
    private _id: number | undefined;
    private _name: string;
    private _sigle: string;
    private _students: number[] = [];
    private _teachers: number[] = [];

    constructor(name: string, sigle: string) {
        this._name = name;
        this._sigle = sigle;
    }

    get id() {
        return this._id as number;
    }

    set id(id: number) {
        this._id = id;
    }

    get name(): string {
        return this._name;
    }

    set name(value: string) {
        this._name = value;
    }

    get sigle(): string {
        return this._sigle;
    }

    set sigle(value: string) {
        this._sigle = value;
    }


    get students(): number[] {
        return this._students;
    }

    set students(value: number[]) {
        this._students = value;
    }

    get teachers(): number[] {
        return this._teachers;
    }

    set teachers(value: number[]) {
        this._teachers = value;
    }

    // toJson() {
    //     return {
    //         _name: this._name,
    //         _sigle: this._sigle,
    //         _students: Array.from(this._students).map(s => ({ id: s.id })), // envoie juste les IDs
    //         _teachers: Array.from(this._teachers).map(t => ({ id: t.id }))
    //     };
    // }
}

File: school-space-services/src/main/java/com/example/schoolspace/model/Course.java
package com.example.schoolspace.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import lombok.*;

import java.util.HashSet;
import java.util.Set;

@Setter
@Getter
@Entity
@Table(name = "courses")
@NoArgsConstructor
@AllArgsConstructor
@ToString
public class Course extends AbstractEntity{

    private String name;

    private String sigle;

    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
    @ManyToMany(fetch = FetchType.LAZY, cascade = CascadeType.MERGE)
    @JoinTable(
            name = "courses_students", // nom de la table d'association
            joinColumns = @JoinColumn(name = "course_id"), // clé étrangère vers Course
            inverseJoinColumns = @JoinColumn(name = "student_id") // clé étrangère vers Student
    )
    private Set<Student> students = new HashSet<>();

    @JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
    @ManyToMany(fetch = FetchType.LAZY, cascade = CascadeType.MERGE)
    @JoinTable(
            name = "courses_teachers", // nom de la table d'association
            joinColumns = @JoinColumn(name = "course_id"), // clé étrangère vers Course
            inverseJoinColumns = @JoinColumn(name = "teachers_id") // clé étrangère vers Student
    )
    @JsonProperty("_teachers")
    private Set<Teacher> teachers = new HashSet<>();

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getSigle() {
        return sigle;
    }

    public void setSigle(String sigle) {
        this.sigle = sigle;
    }

    public Set<Student> getStudents() {
        return students;
    }

    public void setStudents(Set<Student> students) {
        this.students = students;
    }

    public Set<Teacher> getTeachers() {
        return teachers;
    }

    public void setTeachers(Set<Teacher> teachers) {
        this.teachers = teachers;
    }
}
